File: packages/lineage/src/recorder.ts
import { computeSpecHash, type JsonObject, type Surface, type UISpec } from "@kohaku-ui/spec-core";
import type { ComposeTraceLike, Lineage } from "./lineage.js";

/**
 * Adapter conforming to host-rest's ViewRecorder interface.
 * Automatically records the REST surface's compose / events / telemetry into View Lineage.
 */
export interface RestViewRecorder {
  composed(args: {
    spec: UISpec;
    trace: ComposeTraceLike;
    surface: Surface;
    sessionId?: string;
    tenant?: string;
    /** Precomputed hash. If passed, viewComposed does not re-hash the Spec. Unset means internal computation (backward compatible). */
    specHash?: string;
    structureHash?: string;
  }): Promise<void>;
  interacted(args: {
    intentHash: string;
    componentId: string;
    on: string;
    payload: JsonObject;
    surface: Surface;
    sessionId?: string;
    tenant?: string;
  }): Promise<void>;
  rendered?(args: {
    specHash: string;
    surface: Surface;
    renderer: string;
    durationMs?: number;
    tenant?: string;
  }): Promise<void>;
  componentUsed?(args: {
    artifactId: string;
    surface: Surface;
    outcome: "ok" | "error";
    sessionId?: string;
    tenant?: string;
  }): Promise<void>;
  fallback?(args: {
    spec: UISpec;
    reason: string;
    kind: "generation" | "negotiation";
    surface: Surface;
    sessionId?: string;
    tenant?: string;
    /** Precomputed hash. If passed, does not re-hash the Spec. Unset means internal computation (backward compatible). */
    specHash?: string;
  }): Promise<void>;
}

export function createViewRecorder(lineage: Lineage): RestViewRecorder {
  return {
    async composed(args) {
      await lineage.viewComposed(args);
    },
    async interacted(args) {
      await lineage.viewInteracted(args);
    },
    async rendered(args) {
      await lineage.viewRendered(args);
    },
    async componentUsed(args) {
      // Actual-render observation via telemetry. Stamps source so it can be distinguished from the compose-time record (excluded from promotion aggregation).
      await lineage.componentUsed({ ...args, source: "telemetry" });
    },
    async fallback(args) {
      // Derive specHash / intentHash from spec and record view.fallback.
      // Makes the occurrence rate (L1/L2 failures, capability downgrade) observable from lineage.
      // If the caller already computed it (host-rest's deliverComposed shares one computation across
      // recordComposed and recordFallbackIfAny), reuse it instead of hashing the same Spec again.
      const specHash = args.specHash ?? (await computeSpecHash(args.spec));
      await lineage.viewFallback({
        specHash,
        reason: args.reason,
        surface: args.surface,
        kind: args.kind,
        intentHash: args.spec.intent.hash,
        ...(args.sessionId != null ? { sessionId: args.sessionId } : {}),
        ...(args.tenant != null ? { tenant: args.tenant } : {}),
      });
    },
  };
}

File: python/kohaku/src/kohaku/lineage/recorder.py
"""Adapter conforming to host-rest's ViewRecorder (Port of TS packages/lineage/src/recorder.ts).

Auto-records the REST surface's compose / events / telemetry into View Lineage.
computeSpecHash is a synchronous function of kohaku.spec (the existing port's style).
"""

from __future__ import annotations

from kohaku.spec import JsonObject, Surface, UISpec, compute_spec_hash

from .lineage import ComposeTraceLike, Lineage


class ViewRecorder:
    """Equivalent to RestViewRecorder. Wraps Lineage and records the REST surface's events into View Lineage."""

    def __init__(self, lineage: Lineage) -> None:
        self._lineage = lineage

    async def composed(
        self,
        *,
        spec: UISpec,
        trace: ComposeTraceLike,
        surface: Surface,
        session_id: str | None = None,
        tenant: str | None = None,
        spec_hash: str | None = None,
        structure_hash: str | None = None,
    ) -> None:
        await self._lineage.view_composed(
            spec=spec,
            trace=trace,
            surface=surface,
            session_id=session_id,
            tenant=tenant,
            spec_hash=spec_hash,
            structure_hash=structure_hash,
        )

    async def interacted(
        self,
        *,
        intent_hash: str,
        component_id: str,
        on: str,
        payload: JsonObject,
        surface: Surface,
        session_id: str | None = None,
        tenant: str | None = None,
    ) -> None:
        await self._lineage.view_interacted(
            intent_hash=intent_hash,
            component_id=component_id,
            on=on,
            payload=payload,
            surface=surface,
            session_id=session_id,
            tenant=tenant,
        )

    async def rendered(
        self,
        *,
        spec_hash: str,
        surface: Surface,
        renderer: str,
        duration_ms: float | None = None,
        tenant: str | None = None,
    ) -> None:
        await self._lineage.view_rendered(
            spec_hash=spec_hash,
            surface=surface,
            renderer=renderer,
            duration_ms=duration_ms,
            tenant=tenant,
        )

    async def component_used(
        self,
        *,
        artifact_id: str,
        surface: Surface,
        outcome: str = "ok",
        session_id: str | None = None,
        tenant: str | None = None,
    ) -> None:
        # Observed real render via telemetry. Stamp source so it can be distinguished from compose-time recording (excluded from promotion aggregation).
        await self._lineage.component_used(
            artifact_id=artifact_id,
            surface=surface,
            outcome=outcome,
            session_id=session_id,
            source="telemetry",
            tenant=tenant,
        )

    async def fallback(
        self,
        *,
        spec: UISpec,
        reason: str,
        kind: str,
        surface: Surface,
        session_id: str | None = None,
        tenant: str | None = None,
    ) -> None:
        # Derive specHash / intentHash from spec and record view.fallback.
        # Makes the occurrence rate (L1/L2 failure, capability downgrade) observable from lineage.
        spec_hash = compute_spec_hash(spec)
        await self._lineage.view_fallback(
            spec_hash=spec_hash,
            reason=reason,
            surface=surface,
            kind=kind,
            intent_hash=spec.intent.hash,
            session_id=session_id,
            tenant=tenant,
        )


def create_view_recorder(lineage: Lineage) -> ViewRecorder:
    return ViewRecorder(lineage)
